pkg/deployment/reconcile/action_bootstrap_set_password.go

```diff
@@ -0,0 +1,148 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+// Author Adam Janikowski
+//
+
+package reconcile
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/arangodb/go-driver"
+	api "github.com/arangodb/kube-arangodb/pkg/apis/deployment/v1"
+	"github.com/arangodb/kube-arangodb/pkg/util"
+	"github.com/arangodb/kube-arangodb/pkg/util/k8sutil"
+	"github.com/rs/zerolog"
+)
+
+func init() {
+	registerAction(api.ActionTypeBootstrapSetPassword, newBootstrapSetPasswordAction)
+}
+
+func newBootstrapSetPasswordAction(log zerolog.Logger, action api.Action, actionCtx ActionContext) Action {
+	a := &actionBootstrapSetPassword{}
+
+	a.actionImpl = newActionImplDefRef(log, action, actionCtx, defaultTimeout)
+
+	return a
+}
+
+type actionBootstrapSetPassword struct {
+	// actionImpl implement timeout and member id functions
+	actionImpl
+
+	actionEmptyCheckProgress
+}
+
+func (a actionBootstrapSetPassword) Start(ctx context.Context) (bool, error) {
+	spec := a.actionCtx.GetSpec()
+
+	if user, ok := a.action.GetParam("user"); !ok {
+		a.log.Warn().Msgf("User param is not set in action")
+		return true, nil
+	} else {
+		if secret, ok := spec.Bootstrap.PasswordSecretNames[user]; !ok {
+			a.log.Warn().Msgf("User does not exist in password hashes")
+			return true, nil
+		} else {
+			ctx, c := context.WithTimeout(context.Background(), a.Timeout(spec))
+			defer c()
+			if password, err := a.setUserPassword(ctx, user, secret.Get()); err != nil {
+				return false, err
+			} else {
+				passwordSha := util.SHA256FromString(password)
+
+				if err := a.actionCtx.WithStatusUpdate(func(s *api.DeploymentStatus) bool {
+					if s.SecretHashes == nil {
+						s.SecretHashes = &api.SecretHashes{}
+					}
+
+					if s.SecretHashes.Users == nil {
+						s.SecretHashes.Users = map[string]string{}
+					}
+
+					if u, ok := s.SecretHashes.Users[user]; !ok || u != passwordSha {
+						s.SecretHashes.Users[user] = passwordSha
+						return true
+					}
+					return false
+				}); err != nil {
+					return false, err
+				}
+			}
+		}
+	}
+	return true, nil
+}
+
+func (a actionBootstrapSetPassword) setUserPassword(ctx context.Context, user, secret string) (string, error) {
+	a.log.Debug().Msgf("Bootstrapping user %s, secret %s", user, secret)
+
+	client, err := a.actionCtx.GetDatabaseClient(ctx)
+	if err != nil {
+		return "", maskAny(err)
+	}
+
+	password, err := a.ensureUserPasswordSecret(user, secret)
+	if err != nil {
+		return "", maskAny(err)
+	}
+
+	// Obtain the user
+	if u, err := client.User(context.TODO(), user); driver.IsNotFound(err) {
```

Comment:
Why TODO instead of Background ?

Reply:
Will fix, was just used from other place - from k8s utils